File: memoization/grid-traveler.js
// the grid traveler problem that can only move right and down

const gridTraveler = (m, n, memo = {}) => {
  const key = m +','+n;
  if (key in memo) return memo[key];
  if (m ===1 && n === 1) return 1;
  if (m === 0 || n === 0) return 0;
  memo[key] =  gridTraveler(m-1, n, memo) + gridTraveler(m, n-1, memo);
  return memo[key];
}

console.log(gridTraveler(2, 3))
console.log(gridTraveler(4, 5))
console.log(gridTraveler(6, 7))
console.log(gridTraveler(18, 18))

File: memoization/grid_traveler.py
# the grid traveler problem in python

def grid_traveler(m, n, memo = {}):
    key = str(m) + ','+str(n)
    if key in memo: return memo[key]
    if m == 1 and n == 1: return 1;
    if m == 0 or n == 0: return 0;
    memo[key] = grid_traveler(m-1, n, memo) + grid_traveler(m, n-1, memo)
    # print(memo)
    return memo[key]

print(grid_traveler(2, 3))
print(grid_traveler(4, 5))
print(grid_traveler(6, 7))
print(grid_traveler(18, 18))


